### PR TITLE
Fix bugs in ByteProvider and j2k wscript

### PR DESCRIPTION
**Pull request by JonathanMeans**

### CODE REVIEW

File: modules/c++/nitf/source/ByteProvider.cpp [hunk 1]
```diff
@@ -566,11 +562,10 @@ void ByteProvider::getBytes(const void* imageData,
                     imageDataEndRow,
                     startGlobalRowToWrite,
                     imageData,
-                    numPadRowsSoFar,
                     fileOffset,
                     buffers);
 
-            addDES(seg, startRow, buffers);
+            addDES(seg, imageDataEndRow, buffers);
```
Comment (JonathanMeans) [inline modules/c++/nitf/source/ByteProvider.cpp]: `addDES()` checks whether `seg` is the last segment and if the passed row index is the last row of that segment. It will never add the DES bytes if `startRow` is passed.

File: modules/c++/nitf/source/ByteProvider.cpp [hunk 2]
```diff
@@ -432,12 +431,10 @@ void ByteProvider::addImageData(
                 rowsInSegmentSkipped * mNumBytesPerRow;
     }
 
-    const size_t numPadRows = countBytesForImageData(seg, numRowsToWrite,
-            imageDataEndRow) / mNumBytesPerPixel;
-    numRowsToWrite += numPadRows;
-    numPadRowsSoFar += numPadRows;
+    const size_t bytesThisWrite = countBytesForImageData(seg, numRowsToWrite,
```
Comment (JonathanMeans) [inline modules/c++/nitf/source/ByteProvider.cpp]: I think `numPadRows` is a leftover from my botched refactoring. (See line 393). It gives the correct byte count, so we should just use it instead of whatever this what attempting.

File: modules/c++/nitf/source/ByteProvider.cpp [hunk 3]
```diff
@@ -408,16 +408,15 @@ void ByteProvider::addImageData(
...
         nitf::Off& fileOffset,
         NITFBufferList& buffers) const
 {
     const SegmentInfo& imageSegmentInfo = mImageSegmentInfo[seg];
     const size_t segStartRow = imageSegmentInfo.firstRow;
 
     // Figure out what offset of 'imageData' we're writing from
-    const size_t startLocalRowToWrite =
-            startGlobalRowToWrite - startRow + numPadRowsSoFar;
+    // Unless we're reading across segment boundaries, this will be 0
```
Comment (JonathanMeans) [inline modules/c++/nitf/source/ByteProvider.cpp]: This is passing all the tests, but as I stare at it I'm seeing a corner case. I want to see if I can write another test or two that will break for this....

Comment (JonathanMeans) [reply modules/c++/nitf/source/ByteProvider.cpp]: There shouldn't ever be the case where there is padding in between blocks within the same image segment, right? It should all just be at the end? (2500C seems to imply that this is true.) If so, we don't need the pad rows. All the math needed for that happens in `getFileLayout()`.

Comment (asylvest) [reply modules/c++/nitf/source/ByteProvider.cpp]: There won't ever be padding inbetween blocks within an image segment, but there can be padding at the bottom of *each* image segment.  For example, say you had a three segment NITF:
Seg 0: 2000 rows
Seg 1: 4000 rows
Seg 3: 6000 rows
and used 1024 rows / block in each seg.  You'll end up with...

Seg 0: 1024*2 - 2000 = 48 pad rows
Seg 1: 1024*4 - 4000 = 96 pad rows
Seg 2: 1024*6 - 6000 = 144 pad rows

I agree `getFileLayout()` knows how to set the file/header offsets appropriately taking these extra pad rows into account.  I think `numPadRowsSoFar` is getting the pointer arithmetic right.  The input image `imageData` has already been blocked (by `ImageBlocker` or equivalent) which means it itself includes pad rows.  That's what we want so that the correct bytes get written out, but we need to keep track of it as the row the caller is thinking about isn't the same as the offset into that pointer.  For example, let's say that you owned this entire 12,000 row image in memory described above.  After blocking, `imageData` would actually be 2000+48+4000+96+6000+144 = 12,288 rows.  Now say they call `ByteProvider` and it's time to write out rows [2000, 3000) of the image (start of segment 1).  We don't want to go to row 2000 of `imageData` - we need to go to row 2048 of it to actually get the start of segment 1.

If this doesn't make any sense, let me know and I can draw it on the whiteboard.

Comment (JonathanMeans) [reply modules/c++/nitf/source/ByteProvider.cpp]: Looking how `testMultipleWrites()` works in `test_sidd_byte_provider`, `getBytes()` assumes that the data pointer has already been offset appropriately. The problem would come when calculating the offset if you wanted to write rows 1950-2050 across a boundary. So... now I need to figure out why my tests are passing when writing across blocked segments without accounting for this.

**Comment by JonathanMeans** [comment]
This should get things working again. I still don't have a test for this edge case however. It looks like when you go through NITFWriteControl it automatically segments things so that any padding would come at the end. I'm going to have to write a new set of tests in NITRO to get at this.

**Comment by asylvest** [comment]
OK cool - I will try to review shortly.  You could use `nitf::ImageSegmentComputer` except that's what SIX uses under the hood so will result in the same thing in terms of making sure you never get padding crossing segment boundaries.  I do think it's worth adding a test for this in NITRO though as you suggest.

**Comment by asylvest** [comment]
These changes all make sense to me - go ahead and put out PRs to merge this into the other repos now so we get the bug fix and then open a second NITRO PR for the additional NITRO test.

Thanks for tracking this all through!
